Created usdotio python script to add or extract .otio data from USD

ggarra13 opened the pull request:
### Description of Change(s)

Adds a python script to add OpenTimelineIo data to and from a .usd file.   The .otio information is stored in a tree fashion for easy inspecting it with usdview or similar.
You can have multiple .otio timelines, so that, for example, a TV set model carries its own .otio timeline differently from the main .otio at the root.

Its usage is:

usdotio add input.otio sphere.usda   # sphere.usda is modified in place
usdotio save output.otio sphere_with_otio.usda  # output.otio is created from the .otio data found at the root of the usd file.



<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [x ] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [x] I have submitted a signed Contributor License Agreement

ggarra13 committed:
Created usdotio python script to add or extract OpenTimelineIO's
json data from an USD file.

ggarra13 committed:
Merge branch 'PixarAnimationStudios:release' into usdotio

jesschimein commented:
Filed as internal issue #USD-9412

:heavy_exclamation_mark: Please make sure that a signed CLA has been submitted!

jesschimein commented:
/AzurePipelines run

ggarra13 commented:
> ❗ Please make sure that a signed CLA has been submitted!

I did.  I got this email back:

```
redacted
```

FlorianZ commented:
> > ❗ Please make sure that a signed CLA has been submitted!
> 
> I did. I got this email back:

Great, thank you! You're all good then.

ggarra13 commented:
Hi!  How can I get this merged?  It passed all OSes tests but choked on the Azure Pipeline.  Is there anything I can do or help debug it?

FlorianZ commented:
Hi @ggarra13!

We don't automatically merge PRs. Requests typically go through our internal engineering process, which includes code and architectural reviews. This can take a while, considering the volume of requests we receive and the amount of resources we have available.

For plugins, infrastructure, or additions of new tooling - like in your request - we also discuss whether we want to ship the additional code with the OpenUSD repo: Doing so implies that we take on maintenance of the added code, and this requires discussion beyond just the engineering team.

There is typically also a step that involves engaging with the community. For example, for this PR we would want to verify that the schemas align with how the rest of the community expects the data to be represented in USD. Inclusion with the OpenUSD repo implies that we impart an opinion on how this data should be represented across sites as opposed to it being site-specific data. I am not sure if this has happened with your project, just putting it out there as an example. Here is a good place to initiate the community involvement: https://github.com/PixarAnimationStudios/OpenUSD-proposals

Thanks again for your contribution and for offering your help with getting it merged. A good next step would be for you to initiate a high-level conversation about usdotio on https://forum.openusd.org or submit a proposal to https://github.com/PixarAnimationStudios/OpenUSD-proposals. @jesschimein is a great person to reach out to for progress updates.

ggarra13 commented:
> Doing so implies that we take on maintenance of the added code, and this requires discussion beyond just the engineering team.

Sounds reasonable.

> For example, for this PR we would want to verify that the schemas align with how the rest of the community expects the data to be represented in USD.

Again, that sounds fair.  I am curious to know if Pixar or Animal Logic have already added OpenTimelineIO to USD.  NVidia I know currently uses its proprietary Sequencer schema, albeit for a different goal.  And OpenUSD used to support an Audio schema, which it seems it was partially deprecated. 

> Here is a good place to initiate the community involvement: https://github.com/PixarAnimationStudios/OpenUSD-proposals

I will let Michael Davey, who originally hired me to implement this, to comment on his needs for this.  I can open the proposal and discuss it technically, and my thinking of how can this be used in the future.

spiffmon commented:
Just wanted to correct - `UsdMediaSpatialAudio` is not deprecated, and is in use in the Apple ecosystem, at least.  It's more like OpenUSD itself provides no "rendering" capabilities for it, given the difficulty of cross-platform audio and our limited resources.  We are always on the lookout for a good implementation!

ggarra13 commented:
> OpenUSD itself provides no "rendering" capabilities for it, given the difficulty of cross-platform audio and our limited resources. We are always on the lookout for a good implementation!

I am not sure about the requirements of UsdMediaSpatialAudio, but have you guys looked at RtAudio (here on github)?  We are using v5.2 (not sure about the latest one).   @darbyjohnston and myself are using it successfully on Linux, macOS and Windows.  I only found a subtle bug on Linux with heavy multithreading, but was able to work around it.

spiffmon commented:
Thanks for the info - that sounds interesting!

meshula commented:
RtAudio is a stable and reliable choice for sure. UsdMediaSpatialAudio's requirements are very similar to those of classic OpenAL, and RtAudio provides a superset of the requirements of OpenAL. Qt also has a spatial audio module with capabilities sufficient for UsdMediaSpatialAudio (https://doc.qt.io/qt-6/qtspatialaudio-index.html), so that might be the simplest way to enable audio rendering in usdview when it comes right down to it. The missing piece is the equivalent of Hydra for audio, or at least some custom traversal code, to introspect the scenegraph and create and update a runtime environment reflective of the scenegraph state.

ggarra13 committed:
Switched to using IsA() instead of string comparisons.

Started work on 'update' to change Omniverse Sequencer to OTIO.

commit a1e4a3a0aa3f8280837eac732c2d8e5a0d2104cd
Author: Gonzalo Garramuño <ggarra13@gmail.com>
Date:   Thu Feb 22 09:36:52 2024 -0300

    Created usdotio python script to add or extract OpenTimelineIO's
    json data from a codeless Otio's primitive schema.

ggarra13 committed:
Removed personal auxiliary files.

ggarra13 committed:
Update for usdotioupdate.py.

ggarra13 committed:
Minor bug fix.

ggarra13 committed:
WIP usdotioupdate.py.

ggarra13 committed:
Fixed typo.

jesschimein commented:
/AzurePipelines run

md84419 commented:
Where are we at for getting these changes merged, please?
